presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java

```diff
@@ -0,0 +1,461 @@
+
+        cacheFilesToDelete.forEach(LocalRangeCacheManager::tryDeleteFile);
+        return true;
+    }
+
+    private static void tryDeleteFile(Path path)
+    {
+        try {
+            File file = new File(path.toUri());
+            if (file.exists()) {
+                Files.delete(file.toPath());
+            }
+        }
+        catch (IOException e) {
+            // ignore
+        }
+    }
+
+    private static boolean cacheFileEquals(LocalCacheFile left, LocalCacheFile right)
+    {
+        if (left == null && right == null) {
+            return true;
+        }
+
+        if (left == null || right == null) {
+            return false;
+        }
+
+        return left.equals(right);
+    }
+
+    private static class LocalCacheFile
+    {
+        private final long offset;  // the original file offset
+        private final Path path;    // the cache location on disk
+
+        public LocalCacheFile(long offset, Path path)
+        {
+            this.offset = offset;
+            this.path = path;
+        }
+
+        public long getOffset()
+        {
+            return offset;
+        }
+
+        public Path getPath()
+        {
+            return path;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            LocalCacheFile that = (LocalCacheFile) o;
+            return Objects.equals(offset, that.offset) && Objects.equals(path, that.path);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(offset, path);
+        }
+    }
+
+    private static class CacheRange
+    {
+        private final RangeMap<Long, LocalCacheFile> range = TreeRangeMap.create();
+        private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+        public RangeMap<Long, LocalCacheFile> getRange()
+        {
+            return range;
+        }
+
+        public ReadWriteLock getLock()
+        {
+            return lock;
+        }
+    }
+
+    private class CacheRemovalListener
+            implements RemovalListener<Path, Boolean>
+    {
+        @Override
+        public void onRemoval(RemovalNotification<Path, Boolean> notification)
+        {
+            Path path = notification.getKey();
+            CacheRange cacheRange = persisted.remove(path);
+            if (cacheRange == null) {
+                return;
+            }
+
+            cacheRemovalExecutor.submit(() -> {
+                Collection<LocalCacheFile> files;
+                cacheRange.lock.readLock().lock();
+                try {
+                    files = cacheRange.getRange().asMapOfRanges().values();
+                }
+                finally {
+                    cacheRange.lock.readLock().unlock();
+                }
+
+                for (LocalCacheFile file : files) {
```

Comment:
Is it possible that another thread get the result from the cache? If that's the case, before we delete the files, should we get a write lock to not disrupt other reads?

Reply:
That is possible. @jessesleeping's suggestion is to provide a jitter to delay the deletion. My original thought was to create a concurrent hash set as a "safe pool" to protect files being read. But given the chance of getting into this situation is low, we may just fail the read and do it in a simple way.

Reply:
Let's add a comment about this as it is not obvious about the decision we made :) Looks like query won't fail, just a cache miss, which is okay.